boards/BMS/Test/Src/imd_test.cpp

```diff
@@ -0,0 +1,267 @@
+#include "fff.h"
+#include <gtest/gtest.h>
+#include <utility>
+
+extern "C"
+{
+    DEFINE_FFF_GLOBALS;
+#include "imd_fff.h"
+#include "App_Imd.h"
+}
+
+class Imd_Test : public testing::Test
+{
+  protected:
+    void SetUp() override
+    {
+        imd = App_Imd_Create(
+            get_pwm_frequency, 2.0f, get_pwm_duty_cycle,
+            get_seconds_since_power_on);
+
+        RESET_FAKE(get_pwm_frequency);
+        RESET_FAKE(get_pwm_duty_cycle);
+        RESET_FAKE(get_seconds_since_power_on);
+    }
+
+    void TearDown() override
+    {
+        ASSERT_TRUE(imd != NULL);
+        App_Imd_Destroy(imd);
+        imd = NULL;
+    }
+
+    void SetPwmFrequencyTolerance(float tolerance)
+    {
+        ASSERT_TRUE(imd != NULL);
+        App_Imd_Destroy(imd);
+        imd = App_Imd_Create(
+            get_pwm_frequency, tolerance, get_pwm_duty_cycle,
+            get_seconds_since_power_on);
+    }
+
+    void SetImdCondition(enum Imd_Condition condition)
+    {
+        float mapping[NUM_OF_IMD_CONDITIONS] = {
+            [IMD_SHORT_CIRCUIT] = 0.0f,          [IMD_NORMAL] = 10.0f,
+            [IMD_UNDERVOLTAGE_DETECTED] = 20.0f, [IMD_SST] = 30.0f,
+            [IMD_DEVICE_ERROR] = 40.0f,          [IMD_EARTH_FAULT] = 50.0f,
+        };
+
+        get_pwm_frequency_fake.return_val = mapping[condition];
+        App_Imd_Tick(imd);
+        ASSERT_EQ(condition, App_Imd_GetCondition(imd));
+    }
+
+    struct Imd *imd;
+};
+
+TEST_F(
+    Imd_Test,
+    check_insulation_resistance_for_normal_and_undervoltage_conditions)
+{
+    std::vector<enum Imd_Condition> conditions = { IMD_NORMAL,
+                                                   IMD_UNDERVOLTAGE_DETECTED };
+
+    for (auto &condition : conditions)
+    {
+        SetImdCondition(condition);
+
+        // From ISOMETER® IR155-3203/IR155-3204 manual:
+        //     Insulation Resistance =
+        //         (90% * 1200kOhm) / (duty_cycle - 5%) - 1200kOhm
+
+        constexpr float MAX_VALID_DUTY_CYCLE = 95.0f;
```

Comment:
General testing strategy for all the tests: Define the `MIN/MAX` duty cycle at the top. Then exercise `0.1` below `MIN`, `MIN`, `(MIN+MAX)/2`, `MAX`, `0.1` above`MAX`.